.github/guides/ai/LEARN_AI.md

```diff
@@ -0,0 +1,174 @@
+	var/atom/target = weak_target?.resolve()
+	var/datum/targetting_datum/targetting_datum = controller.blackboard[targetting_datum_key]
+
+	if(!targetting_datum.can_attack(basic_mob, target))
+		///We have a target that is no longer valid to attack. Remember that returning doesn't end the behavior, JUST this single performance. So we call "finish_action" with whether it succeeded in doing what it wanted to do (it didn't, so FALSE) and the blackboard keys passed into this behavior.
+		finish_action(controller, FALSE, target_key)
+		return //don't forget to end the performance too
+
+	var/hiding_target = targetting_datum.find_hidden_mobs(basic_mob, target) //If this is valid, theyre hidden in something!
+
+	controller.blackboard[hiding_location_key] = hiding_target
+
+	///and finally, we're in range, we have a valid target, we can attack. When they fall into crit, they will no longer be a valid target, to the melee behavior will end.
+	if(hiding_target) //Slap it!
+		basic_mob.melee_attack(hiding_target)
+	else
+		basic_mob.melee_attack(target)
+
+///and so the action has ended. we can now clean up the AI's blackboard based on the success of the action, and the keys passed in.
+/datum/ai_behavior/basic_melee_attack/finish_action(datum/ai_controller/controller, succeeded, target_key, targetting_datum_key, hiding_location_key)
+	. = ..()
+	///if the behavior failed, the target is no longer valid, so we should lose aggro of them. We remove the target_key (which could be anything, it's whatever key was passed into the behavior by the subtree) from the blackboard. Couldn't do THAT with normal variables!
+	if(!succeeded)
+		controller.blackboard -= target_key
+```
+
+One rule for subtrees: **NO setting blackboard keys in subtrees!** We want blackboard to handle them, subtrees are just for checking the state either via looking at the pawn (atom the AI controls) or looking at the blackboard.
+
+BAD:
+
+```dm
+	if(prob(50))
+		var/list/possible_targets = list()
+		for(var/atom/thing in view(2, living_pawn))
+			if(!thing.mouse_opacity)
+				continue
+			if(thing.IsObscured())
+				continue
+			possible_targets += thing
+		var/atom/target = pick(possible_targets)
+		if(target)
+			controller.blackboard[BB_MONKEY_CURRENT_PRESS_TARGET] = target
+			controller.queue_behavior(/datum/ai_behavior/use_on_object, BB_MONKEY_CURRENT_PRESS_TARGET)
+			return
+```
+
+GOOD:
+
+```dm
+	if(!controller.blackboard[BB_MONKEY_CURRENT_PRESS_TARGET])
+		controller.queue_behavior(/datum/ai_behavior/find_nearby, BB_MONKEY_CURRENT_PRESS_TARGET)
+		return
+
+	if(prob(50))
+		controller.queue_behavior(/datum/ai_behavior/use_on_object, BB_MONKEY_CURRENT_PRESS_TARGET)
+		return SUBTREE_RETURN_FINISH_PLANNING
+```
+As you can see we're putting the search behavior... on a behavior! The monkey will plan from other subtrees while it finds nearby, but now we have a BB any subtree can look at, any behavior can modify or remove.
+
+## Advanced Concepts
```

Comment:
armhulen is so confident in his system that he doesnt think ANY concepts are advanced and did not write anything

Reply:
ROUGH DRAFT I'M THINKIN

Reply:
might put signal hooking into it